opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -204,19 +205,120 @@ def __repr__(self):
+                "timestamp=" + util.ns_to_iso_str(event.timestamp),
+                "attributes="
+                + format_attributes(event.attributes, indent + 2),
+                ")",
+            ]
+
+            for index in range(1, len(text) - 1):
+                text[index] = " " * (2 + indent) + text[index]
+
+            text[0] = " " * indent + text[0]
+            text[-1] = " " * indent + text[-1]
+
+            return os.linesep.join(text)
+
+        def format_events(events, indent=0):
+            if not events:
+                return "None"
+            text = ""
+            for event in events:
+                text += os.linesep + format_event(event, indent)
+            return text
+
+        def format_link(link, indent=0):
+            text = [
+                "Link(",
+                "context="
+                + format_context(link.context, 2 + indent + len("context=")),
+                "attributes=" + format_attributes(link.attributes, indent + 2),
+                ")",
+            ]
+
+            # add indentation
+            for index in range(1, len(text) - 1):
+                text[index] = " " * (2 + indent) + text[index]
+            text[0] = " " * indent + text[0]
+            text[-1] = " " * indent + text[-1]
+
+            return os.linesep.join(text)
+
+        def format_links(links, indent=0):
+            if not links:
+                return "None"
+            text = ""
+            for link in links:
+                text += os.linesep + format_link(link, indent)
+            return text
+
+        parent_id = "None"
+        if self.parent is not None:
+            if isinstance(self.parent, Span):
+                ctx = self.parent.context
+                parent_id = trace_api.format_span_id(ctx.span_id)
+            elif isinstance(self.parent, SpanContext):
+                parent_id = trace_api.format_span_id(self.parent.span_id)
+
+        start_time = "None"
+        if self.start_time:
+            start_time = util.ns_to_iso_str(self.start_time)
+
+        end_time = "None"
+        if self.end_time:
+            end_time = util.ns_to_iso_str(self.end_time)
+
+        text = [
+            type(self).__name__ + "(",
+            'name="' + self.name + '"',
+            "context="
+            + format_context(self.context, 2 + indent + len("context=")),
+            "kind=" + str(self.kind),
+            "parent_id=" + parent_id,
+            "start_time=" + start_time,
+            "end_time=" + end_time,
+            "attributes=" + format_attributes(self.attributes, 2),
+            "events=" + format_events(self.events, indent + 4),
+            "links=" + format_links(self.links, indent + 4),
+            ")",
+        ]
+
+        # add indentation
+        for index in range(1, len(text) - 1):
+            text[index] = " " * (2 + indent) + text[index]
+        text[-1] = " " * indent + text[-1]
+
+        return os.linesep.join(text)
```

Comment:
This is really great, I have also found the current format of the spans to be quite hard to read. :+1:

May I suggest to use an already implemented formatter? This could save a significant amount of code, `json.dumps`, for example:

```python
ipdb> root
Span(name="root", context=SpanContext(trace_id=0x7e1649ba1b796155a512c634af2344e5, span_id=0xe1e66f902a76d737, trace_state={}))
ipdb> print(dumps({key: str(value) for key, value in root.__dict__.items() if not key.startswith("_")}, indent=4))
{
    "name": "root",
    "context": "SpanContext(trace_id=0x7e1649ba1b796155a512c634af2344e5, span_id=0xe1e66f902a76d737, trace_state={})",
    "parent": "None",
    "sampler": "<opentelemetry.trace.sampling.StaticSampler object at 0x7f1b8507a820>",
    "trace_config": "None",
    "resource": "<opentelemetry.sdk.resources.Resource object at 0x7f1b850d7700>",
    "kind": "SpanKind.INTERNAL",
    "span_processor": "<opentelemetry.sdk.trace.MultiSpanProcessor object at 0x7f1b84f4a520>",
    "status": "None",
    "attributes": "BoundedDict({'component': 'http', 'http.method': 'GET', 'http.url': 'https://example.com:779/path/12/?q=d#123', 'http.status_code': 200, 'http.status_text': 'OK', 'misc.pi': 3.14, 'attr-key': 'attr-value2', 'empty-list': [], 'list-of-bools': [True, True, False], 'list-of-numerics': [123, 314, 0]}, maxlen=32)",
    "events": "BoundedList([], maxlen=128)",
    "links": "BoundedList([], maxlen=32)",
    "instrumentation_info": "InstrumentationInfo(tests.trace.test_trace, )"
}
```
I think it makes more sense to define `__str__` methods for our things like `opentelemetry.trace.sampling.StaticSampler` or `opentelemetry.sdk.resources.Resource` so that this can display them correctly and leave the formatting and indentation to an already existing formatter.

Reply:
Yes, I agree with you. I had tried multiple times to get a nice output but I wasn't able. Finally I think I got a good solution, already pushed the change and updated the PR description.